Document that groupByKey will OOM for large keys

RadixSeven opened the pull request:
This pull request is my own work and I license it under Spark's open-source license.

This contribution is an improvement to the documentation. I documented that the maximum number of values per key for groupByKey is limited by available RAM (see [Datablox](http://databricks.gitbooks.io/databricks-spark-knowledge-base/content/best_practices/prefer_reducebykey_over_groupbykey.html) and [the spark mailing list](http://apache-spark-user-list.1001560.n3.nabble.com/Understanding-RDD-GroupBy-OutOfMemory-Exceptions-tp11427p11466.html)).

Just saying that better performance is available is not sufficient. Sometimes you need to do a group-by - your operation needs all the items available in order to complete. This warning explains the problem.

Commit:
Doc that groupByKey will OOM for large keys

Documented that maximum number of values per key for groupByKey is limited by available RAM (see [Datablox][datablox link] and [the spark mailing list][list link]).

Just saying that better performance is available is not sufficient. Sometimes you need to do a group-by - your operation needs all the items available in order to complete. This warning explains the problem.

[datablox link]: http://databricks.gitbooks.io/databricks-spark-knowledge-base/content/best_practices/prefer_reducebykey_over_groupbykey.html
[list link]: http://apache-spark-user-list.1001560.n3.nabble.com/Understanding-RDD-GroupBy-OutOfMemory-Exceptions-tp11427p11466.html

AmplabJenkins commented:
Can one of the admins verify this patch?

rxin commented on core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala:
```diff
@@ -437,6 +437,9 @@ class PairRDDFunctions[K, V](self: RDD[(K, V)])
    * or [[PairRDDFunctions.reduceByKey]] will provide much better performance.
+   *
+   * Note: As currently implemented, GroupByKey must be able to hold all the key-value pairs for any
+   * key in memory. If a key has too many values, it can result in an out-of-memory exception.
```
maybe "out-of-memory exception" => "[[OutOfMemoryError]]"?

srowen commented on core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala:
```diff
@@ -437,6 +437,9 @@ class PairRDDFunctions[K, V](self: RDD[(K, V)])
    * aggregation (such as a sum or average) over each key, using [[PairRDDFunctions.aggregateByKey]]
    * or [[PairRDDFunctions.reduceByKey]] will provide much better performance.
+   *
+   * Note: As currently implemented, GroupByKey must be able to hold all the key-value pairs for any
```
Nit: the method is `groupByKey`

Commit:
groupByKey docs naming updates

Fixed capitalization of groupByKey method and used canonical name of OutOfMemoryError

rxin commented:
Jenkins, test this please.

SparkQA commented:
[Test build #25247 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25247/consoleFull) for   PR 3936 at commit [`5b6f4e9`](https://github.com/apache/spark/commit/5b6f4e96b365641ee028f183f4df6d6f9bfcb702).
 * This patch merges cleanly.

andrewor14 commented:
LGTM, I will merge this to master once the tests pass

SparkQA commented:
[Test build #25247 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25247/consoleFull) for   PR 3936 at commit [`5b6f4e9`](https://github.com/apache/spark/commit/5b6f4e96b365641ee028f183f4df6d6f9bfcb702).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25247/
Test PASSed.
